### PR TITLE
Synchronized output stream

### PR DESCRIPTION
**Pull request by brookedolny**
Currently, this only is used for `std::cout` but could be used for other output streams. `PrintingFailureLogger` and `ContextEventLogger` both use the `sync_ostream_wrapper` for outputting to `std::cout`.

### CODE REVIEW

File: src/Interpreter/PrintingFailureLogger.cpp [hunk 1]
```diff
@@ -78,7 +79,9 @@ void PrintingFailureLogger::log_failure(const Model* model, const Context& ctx,
   ss << "Assertion:\n" << failure.check << '\n';
 
   std::unique_lock lock(mtx);
-  *os << ss.str() << std::flush;
+  sync_ostream_wrapper sync(*os);
+  sync << ss.str();
+  // *os << ss.str() << std::flush;
```
Comment (Phantomical) [inline src/Interpreter/PrintingFailureLogger.cpp:84]: This commented out code should probably just be deleted.

File: include/caffeine/Support/SyncOStream.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  void flush();
+
+public:
+  sync_ostream_wrapper(std::ostream& o);
+  ~sync_ostream_wrapper();
+
+  std::string get_formatted_string();
+
+  template <typename T>
+  sync_ostream_wrapper& operator<<(T&& s) {
```
Comment (Phantomical) [inline include/caffeine/Support/SyncOStream.h:30]: This should take `const T&` since it's not modifying the value.

**Review by Phantomical — APPROVED** [review]
LGTM with two nits :)

File: src/Interpreter/PrintingFailureLogger.cpp (hunk 1 above)
Comment (brookedolny) [reply src/Interpreter/PrintingFailureLogger.cpp]: Deleted

File: include/caffeine/Support/SyncOStream.h (hunk 2 above)
Comment (brookedolny) [reply include/caffeine/Support/SyncOStream.h]: Changed

**Comment by Phantomical** [comment]
Note that you need to rebase this PR since some CI stuff has been updated. Otherwise this should be good to go :)

**Pull request merged**